src/storage/restrict.rs

```diff
@@ -0,0 +1,231 @@
+
+use std::marker::PhantomData;
+use std::ops::DerefMut;
+use std::borrow::{Borrow, BorrowMut};
+use std::fmt;
+
+use hibitset::BitSet;
+
+use storage::MaskedStorage;
+use world::EntityIndex;
+use {Component, DistinctStorage, Entities, Entity, Index, Join, ParJoin, Storage, UnprotectedStorage};
+
+/// Specifies that the `RestrictedStorage` cannot run in parallel.
+pub enum NormalRestriction { }
+/// Specifies that the `RestrictedStorage` can run in parallel.
+pub enum ParallelRestriction { }
+
+/// Similar to a `MaskedStorage` and a `Storage` combined, but restricts usage
+/// to only getting and modifying the components. That means nothing that would
+/// modify the inner bitset so the iteration cannot be invalidated. For example,
+/// no insertion or removal is allowed.
+pub struct RestrictedStorage<'rf, 'st: 'rf, B, T, R, RT>
+    where T: Component,
+          R: Borrow<T::Storage> + 'rf,
+          B: Borrow<BitSet> + 'rf,
+{
+    bitset: B,
+    data: R,
+    entities: &'rf Entities<'st>,
+    phantom: PhantomData<(T, RT)>,
+}
+
+
+unsafe impl<'rf, 'st: 'rf, B, T, R, ParallelRestriction> ParJoin for &'rf RestrictedStorage<'rf, 'st, B, T, R, ParallelRestriction>
```

Comment:
Ah true, I must've done a `s/RT/ParallelRestriction` instead of just replacing the `RestrictedStorage` one.